A pre-commit hook to magically fix whitespace issues.

SkySkimmer opened the pull request:
It's supposed to leave non git-add'ed changes alone (using `git stash` while working).
It detects whether there is a merge, cherry-pick or rebase in progress. If interactive rebase it can work around it, otherwise it prints any whitespace error (`git diff --check`) and accepts the commit as is.

There may be a smarter way to do this, improvements welcome. I'm going to try it out for some time.

SkySkimmer commented:
>I'm going to try it out for some time.

Note however that emacs already eats my whitespace errors so this isn't the best testing in the world.

psteckler commented:
The man page for `git diff` says that whitespace errors issued for the `--check` flag are controlled by the `core.whitespace` configuration. So maybe there should be such a config that works with this hook, so all users see the same errors.

SkySkimmer commented:
.gitattributes should be doing that.

psteckler commented:
> .gitattributes should be doing that.

Right, I'm suggesting that the `core.whitespace` entry for `.gitattributes` could be added to this PR.

Zimmi48 commented:
@SkySkimmer It would probably make sense to suggest using this commit hook in the Contributing guide, wouldn't it?

SkySkimmer commented:
Actually I think using git apply instead of rebase would be more robust so this isn't finished.

SkySkimmer commented:
New version based on `git apply`.
`git apply --cached` is very nice and lets us cleanly keep staged and unstaged changes separate.

Zimmi48 commented:
Is this now ready and sufficiently tested? It would probably be good if this could be merged and #6268 be closed.

SkySkimmer commented:
It's stable unless someone asks for changes. As I said my editor reduces test coverage but it hasn't broken my commits yet. I don't know if anyone else has been trying it out.
The `needs: review` is also still relevant.

SkySkimmer commented:
In #6528 it was suggested to make configure install the hook. I'm not working on that so feel free to push a commit doing it.

JasonGross commented:
I made https://github.com/SkySkimmer/coq/pull/5 to have `./configure` add the precommit hook automatically.

JasonGross commented on dev/tools/pre-commit:
```diff
@@ -0,0 +1,38 @@
+
+# NB: even though we use --cached it seems to exit code 1 with
+# unstaged whitespace errors. That just means they get fixed though.
+if git diff-index --check --cached HEAD --quiet;
```
I think this is the wrong check.  `--quiet` implies `--exit-code`, which means that this command fails when there are any changes whatsoever.  So this is basically just a check that the commit is not empty.  The whitespace fixer is fast enough that this doesn't really hurt, though.

SkySkimmer replied:
>I think this is the wrong check. --quiet implies --exit-code, which means that this command fails when there are any changes whatsoever. So this is basically just a check that the commit is not empty.

I guess that explains what I saw in the comment.

SkySkimmer commented:
Does anyone know if commit hooks work without issues on windows?

Zimmi48 commented on CONTRIBUTING.md:
```diff
@@ -30,6 +30,8 @@ It's helpful to run the Coq test suite with `make test-suite` before submitting
 Don't be alarmed if the pull request process takes some time. It can take a few days to get feedback, approval on the final changes, and then a merge. Coq doesn't release new versions very frequently so it can take a few months for your change to land in a released version. That said, you can start using the latest Coq `master` branch to take advantage of all the new features, improvements, and f ...
 
+Whitespace discipline (do not indent using tabs, no trailing spaces) is checked by Travis (using `git diff --check`). You can ensure that your commits do not introduce errors by adding the [`dev/tools/pre-commit`](/dev/tools/pre-commit) script to your `.git/hooks/` directory.
+
```
Maybe add "just running configure" will add this, if you do not have a pre-existing pre-commit hook.

Zimmi48 commented on dev/tools/check-eof-newline.sh:
```diff
@@ -1,6 +1,6 @@
 #!/usr/bin/env bash
 
-if [ -z "$(tail -c 1 "$1")" ]
+if ! dev/tools/should-check-whitespace.sh "$1" || [ -z "$(tail -c 1 "$1")" ]
```
Didn't you remove this file in another PR of yours?

SkySkimmer replied:
>Didn't you remove this file in another PR of yours?

#6568 which isn't merged yet.
This PR and that one conflict so I'll just fix whichever gets merged second.

SkySkimmer commented:
Rebased.
I merged the `fix-eof-newline` and `check-eof-newline` by adding a `--fix` argument to the later.

Zimmi48 commented on dev/tools/pre-commit:
```diff
@@ -0,0 +1,47 @@
+#!/bin/sh
+
+# Copy to .git/hooks/ to use.
```
Maybe update this comment.

SkySkimmer replied:
>Maybe update this comment.

Done.

Zimmi48 commented on dev/tools/pre-commit:
```diff
@@ -0,0 +1,47 @@
+then
+    2>&1 echo "Some files had newline errors; they have been fixed in the working tree."
+    2>&1 echo "Make sure to add them before committing."
+    2>&1 echo "This may fix itself if you were using git commit -a, and you try again."
```
Not fully clear what is meant here.

SkySkimmer replied:
>Not fully clear what is meant here.

The errors were fixed only in the worktree, so there was a missing `git add`.
Some thought had to be used to do that `git add` without adding other unstaged changes.
This is now fixed.

Zimmi48 commented on dev/tools/pre-commit:
```diff
@@ -0,0 +1,47 @@
+    # Check that we did fix whitespace
+    if git diff-index --check --cached HEAD;
+    then
+        :
```
Does all POSIX shell have `!`? In that case, you could also just invert the logic of the conditional.

SkySkimmer replied:
>Does all POSIX shell have `!`? In that case, you could also just invert the logic of the conditional.

Done.

SkySkimmer commented:
Someone check on that last commit, then we should be good to go.

mattam82 commented:
When trying to commit something which needs ws-fixing I do: stage, commit in magit, write a log message and then I get:
  1 /usr/local/bin/git … commit --
Auto fixing whitespace issues...
/var/folders/v4/4jfq8t4x4jnb_fs99j19v3300000gn/T/tmp.R92fG2Dv:10: trailing whitespace.

error: kernel/univ.ml: does not match index

That makes some sense because indeed the file has changed. The worst part is that the state after this has lost my changes. I would suggest making the script resistant to these kind of failures and save the users work somewhere first.

SkySkimmer commented:
Technically changes aren't lost, they're in `tmp.R92fG2Dv` (autogenerated name).

Working on it.

SkySkimmer commented:
OK I somehow completely misunderstood what `apply --index` did. It should be fixed now.

mattam82 commented:
If you make a patch that just introduces whitespace at the end of a line:
Auto fixing whitespace issues...
Patches are saved in 'git-fix-ws-index.VSg2r', 'git-fix-ws-index-fixed.Yu3Bp' and 'git-fix-ws-tree.XEcNj'.
If an error destroyes your changes you can recover using them.
(The files are cleaned up on success.)
git-fix-ws-index.VSg2r:10: trailing whitespace.
let compare_levels = Level.compare  
warning: 1 line applied after fixing whitespace errors.
git-fix-ws-index.VSg2r:10: trailing whitespace.
let compare_levels = Level.compare  
warning: 1 line applied after fixing whitespace errors.
error: unrecognized input

mattam82 commented:
The files are there with the (useless in this case) data. Otherwise no problem found by testing this a bit.

mattam82 commented:
Destroyed or destroys BTW

SkySkimmer commented:
>Destroyed or destroys BTW

Fixed, also added commits:
- pre-commit: fail gracefully if fixing whitespace removes all changes
- pre-commit: nicer messages

Zimmi48 commented:
Where are we with this pre-commit hook? It would be great to merge because the linter continues to bother people and if we don't fix this soon, they will again try to rebel to have the linter removed (cf. https://github.com/coq/wiki/commit/4fecbe1712284101e745b97f80fdf4c92baf044b).

SkySkimmer commented:
needs: review for 8cefdd3

Zimmi48 commented:
This particular commit LGTM. If you want some more testing, I can do that.

SkySkimmer commented:
OK then.
